fix salt subset in orchestrator

gtmanfred opened the pull request:
### What does this PR do?
This was originally supposed to be part of https://github.com/saltstack/salt/pull/43933

But went missing for some reason, possibly in my rebase to add the comment about popping expect_minions

### What issues does this PR fix or reference?
Fixes #46456

### Tests?

Yes

### Commits signed with GPG?

Yes

gtmanfred committed:
fix salt subset in orchestrator

gtmanfred committed:
add test for subset in orchestration

cachedout commented on salt/modules/saltutil.py:
```diff
@@ -1273,7 +1273,7 @@ def _exec(client, tgt, fun, arg, timeout, tgt_type, ret, kwarg, **kwargs):
         cmd_kwargs = {
             'tgt': tgt, 'fun': fun, 'arg': arg, 'tgt_type': tgt_type,
-            'ret': ret, 'kwarg': kwarg, 'sub': kwargs['subset'],
+            'ret': ret, 'cli': True, 'kwarg': kwarg, 'sub': kwargs['subset'],
```
I don't understand why this is the correct fix. Could you explain in more detail?

gtmanfred replied:
This is how subset was originally implemented for the `salt` command line

https://github.com/saltstack/salt/commit/27cc8295e69

So calling it from the saltutil command options need to call it the same way.

https://github.com/saltstack/salt/blob/2017.7/salt/cli/salt.py#L144

cachedout replied:
Can we either refactor this then so there is a single source of truth for this behavior or heavily document in both places why this behavior is necessary? When we have these mystery options being passed with no clear documentation on why they are necessary or what they do, it feels a little tech debt-y to me. :]

gtmanfred replied:
:+1:

rallytime commented:
FYI @gtmanfred - just a bump on this one. :)

gtmanfred committed:
document `cli` option for cmd_subset

It would be really nice to have a better name for `cmd_cli` in a future
release, possibly cmd_gen.  I don't feel comfortable making this change in
2017.7 though

gtmanfred commented on salt/client/__init__.py:
```diff
@@ -481,6 +481,8 @@ def cmd_subset(
         following exceptions.
 
         :param sub: The number of systems to execute on
+        :param cli: When this is set to True, a generator is returned,
```
It isn't, but I think we should change it in develop and deprecate it going forward from there.

I can make an issue for that if you want?

gtmanfred committed:
fix pylint
